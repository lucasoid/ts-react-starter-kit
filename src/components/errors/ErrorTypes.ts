export enum ErrorTypes {
    'Error' = 'Error',
    'E404' = 'E404',
    'E500' = 'E500',
}

export class E404 extends Error {
    constructor(message) {
        super(message);
        this.name = ErrorTypes.E404;
    }
}

export class E500 extends Error {
    constructor(message) {
        super(message);
        this.name = ErrorTypes.E500;
    }
}
