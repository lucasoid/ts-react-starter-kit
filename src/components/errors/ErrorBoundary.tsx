import * as React from 'react';
import { ErrorMessage } from './ErrorMessage';
import { ErrorTypes } from './ErrorTypes';

interface IErrorBoundaryProps {
    children?: any;
}

interface IErrorBoundaryState {
    hasError: boolean;
    errorName: string;
}

export class ErrorBoundary extends React.Component<IErrorBoundaryProps, IErrorBoundaryState> {
    state = {
        hasError: false,
        errorName: '',
    };

    constructor(props) {
        super(props);
    }

    static getDerivedStateFromError(error) {
        let name = ErrorTypes[error.name] || ErrorTypes.Error;
        return { hasError: true, errorName: name };
    }

    componentDidCatch(error, errorInfo) {
        console.error(error, errorInfo);
    }

    render() {
        if (this.state.hasError) {
            switch (this.state.errorName) {
                case ErrorTypes.E404:
                    return <ErrorMessage>⚠ Not found</ErrorMessage>;
                case ErrorTypes.E500:
                case ErrorTypes.Error:
                default:
                    return <ErrorMessage>⚠ Something went wrong</ErrorMessage>;
            }
        } else return this.props.children;
    }
}
