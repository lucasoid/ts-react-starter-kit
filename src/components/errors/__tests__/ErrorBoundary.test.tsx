import * as React from 'react';
import { mount } from 'enzyme';
import { ErrorBoundary } from '../ErrorBoundary';
import { E404, E500 } from '../ErrorTypes';

describe('ErrorBoundary', () => {
    it('displays a Not Found error if the error is of type E404', () => {
        let Child: React.FC<{}> = () => {
            throw new E404('is a 404');
        };
        let wrapper = mount(
            <ErrorBoundary>
                <Child />
            </ErrorBoundary>,
        );
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.text().toLowerCase()).toContain('not found');
    });

    it('displays a 500 error if the error is of type E500 or anything else', () => {
        let Child: React.FC<{}> = () => {
            throw new E500('is a 500');
        };
        let wrapper = mount(
            <ErrorBoundary>
                <Child />
            </ErrorBoundary>,
        );
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.text().toLowerCase()).toContain('something went wrong');

        class CustomError extends Error {
            constructor(msg) {
                super(msg);
                this.name = 'CustomError';
            }
        }
        Child = () => {
            throw new CustomError('is a custom error');
        };
        wrapper = mount(
            <ErrorBoundary>
                <Child />
            </ErrorBoundary>,
        );
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.text().toLowerCase()).toContain('something went wrong');
    });

    it('renders the child tree if no errors are caught', () => {
        let Child: React.FC<{}> = () => <div>OK</div>;
        let wrapper = mount(
            <ErrorBoundary>
                <Child />
            </ErrorBoundary>,
        );
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.text().toLowerCase()).toContain('ok');
    });
});
