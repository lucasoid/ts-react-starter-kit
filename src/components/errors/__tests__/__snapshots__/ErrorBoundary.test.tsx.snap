// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorBoundary displays a 500 error if the error is of type E500 or anything else 1`] = `
<ErrorBoundary>
  <ErrorMessage>
    <div
      className="errorMessage light"
    >
      ⚠ Something went wrong
    </div>
  </ErrorMessage>
</ErrorBoundary>
`;

exports[`ErrorBoundary displays a 500 error if the error is of type E500 or anything else 2`] = `
<ErrorBoundary>
  <ErrorMessage>
    <div
      className="errorMessage light"
    >
      ⚠ Something went wrong
    </div>
  </ErrorMessage>
</ErrorBoundary>
`;

exports[`ErrorBoundary displays a Not Found error if the error is of type E404 1`] = `
<ErrorBoundary>
  <ErrorMessage>
    <div
      className="errorMessage light"
    >
      ⚠ Not found
    </div>
  </ErrorMessage>
</ErrorBoundary>
`;

exports[`ErrorBoundary renders the child tree if no errors are caught 1`] = `
<ErrorBoundary>
  <Child>
    <div>
      OK
    </div>
  </Child>
</ErrorBoundary>
`;
